llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -20130,6 +20105,135 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
   return Subtarget.getMinimumJumpTableEntries();
 }
 
+void RVVArgDispatcher::constructHelper(Type *Ty) {
+  const DataLayout &DL = MF->getDataLayout();
+  const Function &F = MF->getFunction();
+  LLVMContext &Context = F.getContext();
+
+  StructType *STy = dyn_cast<StructType>(Ty);
+  if (STy && STy->containsHomogeneousScalableVectorTypes()) {
+    Type *ElemTy = STy->getTypeAtIndex(0U);
+    EVT VT = TLI->getValueType(DL, ElemTy);
+    MVT RegisterVT =
+        TLI->getRegisterTypeForCallingConv(Context, F.getCallingConv(), VT);
+
+    RVVArgInfos.push_back({STy->getNumElements(), RegisterVT, false});
+  } else {
+    SmallVector<EVT, 4> ValueVTs;
+    ComputeValueVTs(*TLI, DL, Ty, ValueVTs);
+
+    for (unsigned Value = 0, NumValues = ValueVTs.size(); Value != NumValues;
+         ++Value) {
+      EVT VT = ValueVTs[Value];
+      MVT RegisterVT =
+          TLI->getRegisterTypeForCallingConv(Context, F.getCallingConv(), VT);
+      unsigned NumRegs =
+          TLI->getNumRegistersForCallingConv(Context, F.getCallingConv(), VT);
+
+      // Skip non-RVV register type
+      if (!RegisterVT.isVector())
+        continue;
+
+      if (RegisterVT.isFixedLengthVector())
+        RegisterVT = TLI->getContainerForFixedLengthVector(RegisterVT);
+
+      RVVArgInfo Info{1, RegisterVT, false};
+
+      while (NumRegs--)
+        RVVArgInfos.push_back(Info);
+    }
+  }
+}
+
+void RVVArgDispatcher::construct(bool IsRet) {
+  const Function &F = MF->getFunction();
+
+  if (IsRet)
+    constructHelper(F.getReturnType());
```

Comment:
Maybe we can have 4 constructors of `RVVArgDispatcher` that take the return type or the correctly typed list? Then have each place we construct pass their list. Seems like it might be cleaner.

Reply:
Sure~